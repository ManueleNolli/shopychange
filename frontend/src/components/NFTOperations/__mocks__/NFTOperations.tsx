import { divWithoutChildrenMock } from '../../../../mocksUtils'

const NFTOperationsMock = () => divWithoutChildrenMock('NFTOperations')
export default NFTOperationsMock
