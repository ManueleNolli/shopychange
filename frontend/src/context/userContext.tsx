import React, { createContext, useContext, useState, ReactNode } from 'react'

interface UserContextType {
  userAddress: string | null
  blockchainNetworkId: number | null
  updateUserContext: (address: string, network: number) => void
  clearUserContext: () => void
}

const UserContext = createContext<UserContextType | undefined>(undefined)

export function UserProvider({ children }: { children: ReactNode }) {
  const [userAddress, setUserAddress] = useState<string | null>(null)
  const [blockchainNetworkId, setBlockchainNetworkId] = useState<number | null>(
    null
  )

  const updateUserContext = (address: string, networkId: number) => {
    setUserAddress(address)
    setBlockchainNetworkId(networkId)
  }

  const clearUserContext = () => {
    setUserAddress(null)
    setBlockchainNetworkId(null)
  }

  return (
    <UserContext.Provider
      value={{
        userAddress,
        blockchainNetworkId,
        updateUserContext,
        clearUserContext,
      }}
    >
      {children}
    </UserContext.Provider>
  )
}

export function useUserContext(): UserContextType {
  const context = useContext(UserContext)
  if (!context) {
    throw new Error('useUserContext must be used within a UserProvider')
  }
  return context
}
