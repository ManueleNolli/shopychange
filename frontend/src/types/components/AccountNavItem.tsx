import { ReactElement } from 'react'

export type AccountNavItem = {
  label: string
  icon?: ReactElement
  href?: string
  action?: () => void
}
