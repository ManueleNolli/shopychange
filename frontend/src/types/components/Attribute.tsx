export type Attribute = {
  traitType: string
  value: string
}
