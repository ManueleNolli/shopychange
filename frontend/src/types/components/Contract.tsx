export type Contract = {
  name: string
  address: string
  symbol: string
}
