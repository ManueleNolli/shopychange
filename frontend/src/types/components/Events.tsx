export enum EventType {
  Mint = 'EventMint',
  Burn = 'EventBurn',
  Transfer = 'EventTransfer',
  Approval = 'EventApproval',
  ApprovalForAll = 'EventApprovalForAll',
  SaleCreated = 'EventSaleCreated',
  SalePriceModified = 'EventSalePriceModified',
  SaleCancelled = 'EventSaleCancelled',
  SaleBought = 'EventSaleBought',
}

export interface EventGeneric {
  date: string
  __typename: EventType
}

export interface EventTransfer extends EventGeneric {
  fromAddress: string
  toAddress: string
}

export interface EventApproval extends EventGeneric {
  owner: string
  approved: string
}

export interface EventApprovalForAll extends EventGeneric {
  owner: string
  operator: string
  isApproved: boolean
}

export interface EventMint extends EventGeneric {
  toAddress: string
}

export interface EventBurn extends EventGeneric {
  fromAddress: string
}

export interface EventSaleCreated extends EventGeneric {
  seller: string
  price: number
}

export interface EventSalePriceModified extends EventGeneric {
  seller: string
  previousPrice: number
  price: number
}

export interface EventSaleCancelled extends EventGeneric {
  seller: string
}

export interface EventSaleBought extends EventGeneric {
  seller: string
  buyer: string
  price: number
}

export type EventNFT =
  | EventTransfer
  | EventApproval
  | EventApprovalForAll
  | EventMint
  | EventBurn
  | EventSaleCreated
  | EventSalePriceModified
  | EventSaleCancelled
  | EventSaleBought

export type NFTHistory = EventNFT[]
