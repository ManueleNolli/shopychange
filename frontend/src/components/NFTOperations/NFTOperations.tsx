import React from 'react'
import { Center, Divider, Flex, HStack } from '@chakra-ui/react'

import { SaleStatus } from '../../types/components/SaleStatus'
import { NFTOperationsProps } from '../../types/props/NFTOperationsProps'

import CancelSaleManager from '../CancelSale/CancelSaleManager/CancelSaleManager'
import ModifySaleManager from '../ModifySale/ModifySaleManager/ModifySaleManager'
import BuyManager from '../Buy/BuyManager/BuyManager'

import SellOrIsForSale from '../Sell/SellOrIsForSale/SellOrIsForSale'
import { colors } from '../../styles/Colors'
import CollectionOwnerCheck from '../../utils/CollectionOwnerCheck/CollectionOwnerCheck'
import ShadowButton from '../ShadowButton/ShadowButton'
import WithdrawRoyaltyButton from '../WithdrawRoyaltyButton/WithdrawRoyaltyButton'

import { useNavigate } from 'react-router-dom'
import { useUserContext } from '../../context/userContext'

/*
Render different buttons for different roles:
    - Owner: Sell, Modify, Cancel
    - possible Buyer: Buy
    - other: nothing
*/

export default function NFTOperations({
  nft,
  sale,
  owner,
  onUpdate,
}: NFTOperationsProps) {
  const { userAddress } = useUserContext()
  const navigate = useNavigate()

  const renderOtherActions = () => {
    return (
      <Flex>
        <HStack spacing="2vw">
          <CollectionOwnerCheck
            collectionAddress={nft.contractAddress}
            redirectHome={false}
          >
            <ShadowButton
              label={'Modify Token Royalties'}
              bgColor={`${colors.primaryChakraColor}.400`}
              textColor={'white'}
              shadowColor={colors.primaryShadowColor}
              hoverColor={`${colors.primaryChakraColor}.500`}
              focusColor={`${colors.primaryChakraColor}.500`}
              onClick={() =>
                navigate(`/${nft.contractAddress}/${nft.tokenId}/royalties`)
              }
            />
          </CollectionOwnerCheck>
          <WithdrawRoyaltyButton
            collectionAddress={nft.contractAddress}
            tokenId={nft.tokenId}
          />
        </HStack>
      </Flex>
    )
  }

  if (owner === userAddress) {
    return (
      <Center flexDirection={'column'}>
        <Flex>
          <HStack spacing="2vw">
            <SellOrIsForSale nft={nft} onUpdate={onUpdate} sale={sale} />

            {sale?.status === SaleStatus.LISTED ? (
              <>
                <ModifySaleManager
                  nft={nft}
                  actualPrice={sale.price}
                  onUpdate={onUpdate}
                />
                <CancelSaleManager nft={nft} onUpdate={onUpdate} />
              </>
            ) : null}
          </HStack>
        </Flex>
        <Divider my="2vh" w="50vw" />
        {renderOtherActions()}
      </Center>
    )
  } else if (userAddress && sale?.status === SaleStatus.LISTED) {
    return (
      <Center flexDirection={'column'}>
        <BuyManager nft={nft} price={sale.price} onUpdate={onUpdate} />
        <Divider my="2vh" w="50vw" />
        {renderOtherActions()}
      </Center>
    )
  } else {
    return <Center>{renderOtherActions()}</Center>
  }
}
