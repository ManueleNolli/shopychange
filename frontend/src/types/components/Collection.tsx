import { Contract } from './Contract'
import { NFT } from './NFT'

export type Collection = {
  contract: Contract
  nfts: NFT[]
}
