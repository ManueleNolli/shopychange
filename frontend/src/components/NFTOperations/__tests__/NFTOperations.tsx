import React from 'react'
import { render, act } from '@testing-library/react'
import NFTOperations from '../NFTOperations'
import { SaleStatus } from '../../../types/components/SaleStatus'
import { useUserContext } from '../../../context/userContext'

jest.mock('../../Sell/SellOrIsForSale/SellOrIsForSale')

jest.mock('../../../utils/CollectionOwnerCheck/CollectionOwnerCheck', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="CollectionOwnerCheck">{children}</div>
  ),
}))

jest.mock('../../WithdrawRoyaltyButton/WithdrawRoyaltyButton')
jest.mock('../../CancelSale/CancelSaleManager/CancelSaleManager')
jest.mock('../../ModifySale/ModifySaleManager/ModifySaleManager')
jest.mock('../../Buy/BuyManager/BuyManager')
jest.mock('../../../context/userContext', () => ({
  useUserContext: jest.fn(),
}))

const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}))

const mockProps = {
  nft: {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
    name: 'Test NFT',
  },
  sale: {
    nftData: {
      contractAddress: '0x0000000000',
      tokenId: 0,
      image:
        'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q',
      name: 'Test NFT',
    },
    seller: '0x12345',
    price: 0,
    status: SaleStatus.LISTED,
  },
  owner: '0x12345',
  onUpdate: jest.fn(),
}
describe('NFTOperations', () => {
  test('renders correctly, owner and listed', async () => {
    ;(useUserContext as jest.Mock).mockReturnValue({
      userAddress: '0x12345',
    })

    const tree = render(<NFTOperations {...mockProps} />)
    await act(async () => {
      await Promise.resolve() // Wait for state update
    })

    expect(tree.container).toMatchSnapshot()

    // owner an listed
    expect(tree.getByTestId('SellOrIsForSale')).toBeInTheDocument()
    expect(tree.getByTestId('CancelSaleManager')).toBeInTheDocument()
    expect(tree.getByTestId('ModifySaleManager')).toBeInTheDocument()
    expect(tree.queryByTestId('BuyManager')).not.toBeInTheDocument()
  })

  test('renders correctly, possible buyer', async () => {
    ;(useUserContext as jest.Mock).mockReturnValue({
      userAddress: '0x54321',
    })

    const tree = render(<NFTOperations {...mockProps} />)
    await act(async () => {
      await Promise.resolve() // Wait for state update
    })

    expect(tree.container).toMatchSnapshot()

    // possible buyer
    expect(tree.getByTestId('BuyManager')).toBeInTheDocument()
    expect(tree.queryByTestId('SellOrIsForSale')).not.toBeInTheDocument()
    expect(tree.queryByTestId('CancelSaleManager')).not.toBeInTheDocument()
    expect(tree.queryByTestId('ModifySaleManager')).not.toBeInTheDocument()
  })

  test('renders correctly, account disconnected', async () => {
    ;(useUserContext as jest.Mock).mockReturnValue({
      userAddress: null,
    })

    const tree = render(<NFTOperations {...mockProps} />)
    await act(async () => {
      await Promise.resolve() // Wait for state update
    })

    expect(tree.container).toMatchSnapshot()

    // possible buyer
    expect(tree.queryByTestId('BuyManager')).not.toBeInTheDocument()
    expect(tree.queryByTestId('SellOrIsForSale')).not.toBeInTheDocument()
    expect(tree.queryByTestId('CancelSaleManager')).not.toBeInTheDocument()
    expect(tree.queryByTestId('ModifySaleManager')).not.toBeInTheDocument()
  })

  test('calls navigate when clicking on Modify Token Royalties', async () => {
    ;(useUserContext as jest.Mock).mockReturnValue({
      userAddress: '0x12345',
    })

    const tree = render(<NFTOperations {...mockProps} />)
    await act(async () => {
      await Promise.resolve() // Wait for state update
    })

    const modifyTokenRoyalties = tree.getByText('Modify Token Royalties')
    expect(modifyTokenRoyalties).toBeInTheDocument()
    modifyTokenRoyalties.click()
    expect(mockNavigate).toHaveBeenCalledWith('/0x0000000000/0/royalties')
  })
})
