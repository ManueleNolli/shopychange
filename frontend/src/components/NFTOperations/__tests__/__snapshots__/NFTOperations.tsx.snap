// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTOperations renders correctly, account disconnected 1`] = `
<div>
  <div
    class="css-gmuwbf"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-stack css-y9ykjr"
      >
        <div
          data-testid="CollectionOwnerCheck"
        >
          <button
            class="chakra-button css-ha7ah3"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Modify Token Royalties
          </button>
        </div>
        <div
          data-testid="WithdrawRoyaltyButton"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`NFTOperations renders correctly, owner and listed 1`] = `
<div>
  <div
    class="css-b93e9j"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-stack css-y9ykjr"
      >
        <div
          data-testid="SellOrIsForSale"
        />
        <div
          data-testid="ModifySaleManager"
        />
        <div
          data-testid="CancelSaleManager"
        />
      </div>
    </div>
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-1r66nag"
    />
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-stack css-y9ykjr"
      >
        <div
          data-testid="CollectionOwnerCheck"
        >
          <button
            class="chakra-button css-ha7ah3"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Modify Token Royalties
          </button>
        </div>
        <div
          data-testid="WithdrawRoyaltyButton"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`NFTOperations renders correctly, possible buyer 1`] = `
<div>
  <div
    class="css-b93e9j"
  >
    <div
      data-testid="BuyManager"
    />
    <hr
      aria-orientation="horizontal"
      class="chakra-divider css-1r66nag"
    />
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-stack css-y9ykjr"
      >
        <div
          data-testid="CollectionOwnerCheck"
        >
          <button
            class="chakra-button css-ha7ah3"
            data-testid="onClick-button"
            style="white-space: initial;"
            type="button"
          >
            Modify Token Royalties
          </button>
        </div>
        <div
          data-testid="WithdrawRoyaltyButton"
        />
      </div>
    </div>
  </div>
</div>
`;
